src/main/java/com/vinylteam/vinyl/web/servlets/SignUpServlet.java

```diff
@@ -52,28 +55,32 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
             response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
             log.debug("Set response status to {'status':{}}", HttpServletResponse.SC_BAD_REQUEST);
             attributes.put("message", "Sorry, the password is empty!");
+            PageGenerator.getInstance().process("registration", attributes, response.getWriter());
         } else {
             if (!password.equals(confirmPassword)) {
                 response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
                 log.debug("Set response status to {'status':{}}", HttpServletResponse.SC_BAD_REQUEST);
                 attributes.put("message", "Sorry, the passwords don't match!");
+                PageGenerator.getInstance().process("registration", attributes, response.getWriter());
             } else {
                 boolean isAdded = userService.add(email, password, discogsUserName);
                 log.debug("Got result of adding user with " +
                         "passed email and password to db {'email':{}, 'isAdded':{}}", email, isAdded);
                 if (isAdded) {
                     response.setStatus(HttpServletResponse.SC_SEE_OTHER);
                     log.debug("Set response status to {'status':{}}", HttpServletResponse.SC_SEE_OTHER);
-                    attributes.put("message", "Please confirm your registration. To do this, follow the link that we sent you by email - " + email);
+                    attributes.put("message", "Please confirm your registration. To do this, follow the link that we sent on your email");
                     attributes.remove("email");
                     attributes.remove("discogsUserName");
+                    confirmationService.sendMessageWithLinkToUserEmail(userService.findByEmail(email).get());
+                    PageGenerator.getInstance().process("confirmation-directions", attributes, response.getWriter());
```

Comment:
one servlet - one service call, I think sending email should be added not to servlet, but into service - userService.add(email, password, discogsUserName), perhaps it can be done even easier (without findByEmail)

Reply:
Are you sure?
How can you have one servlet and one service call when you need to find data for doGet AND write data for doPost?

Reply:
sorry, I mean one method

Reply:
I still need to find data, like email or token, because I need to put it in templates as values. So in doPost I call a method to find data to check if this is correct link, then if it is, find user by that token, then call method to check if login is successful, then call method to update user, then call method to delete token.
How can I break this doPost up?

Reply:
this is SignUpServlet - registration user, am I right?
put confirmationService.sendMessageWithLinkToUserEmail(userService.findByEmail(email).get()); inside userService.add(email, password, discogsUserName)
link always correct here as it is just generated

adding user and adding link can be done in one transaction (if user added, then link also added), when you add user it is possible to return added id from jdbc layer

regarding template you just added (as I can see)
attributes.put("message", "Please confirm your registration. To do this, follow the link that we sent on your email");